### PR TITLE
Only clean up editors on close not on Panel dispose

### PR DESCRIPTION
**Pull request by IanMatthewHuff**
For #4786 

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/datascience/notebook/notebookEditorProvider.ts [hunk 1]
```diff
@@ -184,7 +184,6 @@ export class NotebookEditorProvider implements INotebookEditorProvider {
             return;
         }
         this.trackedVSCodeNotebookEditors.add(editor);
-        this.disposables.push(editor.onDidDispose(() => this.onDidDisposeVSCodeNotebookEditor(editor)));
```
Comment (IanMatthewHuff) [inline src/client/datascience/notebook/notebookEditorProvider.ts:187]: I'm trying to make sure that this isn't going to open up cases where we don't dispose. But this was causing some large issues. We remove editors from the editor list either on OnClose or OnDidDispose. But OnDidDispose as per the API is just for the Panel being disposed. This is triggered by something like dragging a notebook editor into a new panel. So just by dragging to a new panel we'll clean up the document and remove it from the editors list. But OnClose / OnOpen is not called on a drag operation like that.

Currently I don't see any cases where we would not get the OnClose event and we just do our cleanup then like we normally would. onDidCloseNotebookDocument calls the same disposeResourceRelatedToNotebookEditor that this function did.

Also this was only on changing the active document. I would think that if we needed to actually catch all disposes than this would need to be onDidOpenNotebookDocument.

**Review by rchiodo — APPROVED** [review]

**Comment by rchiodo** [comment]
Good catch. Must have been difficult to find.

**Review by DonJayamanne — APPROVED** [review]

**Review by joyceerhl — APPROVED** [review]

File: src/client/datascience/notebook/notebookEditor.ts [hunk 2]
```diff
@@ -97,7 +97,7 @@ export class NotebookEditor implements INotebookEditor {
                 }
             })
         );
-        disposables.push(model.onDidDispose(this._closed.fire.bind(this._closed, this)));
+        disposables.push(model.onDidDispose(this.dispose.bind(this)));
```
Comment (IanMatthewHuff) [inline src/client/datascience/notebook/notebookEditor.ts:100]: This is the change that I made after checking out how this works shutting down VS. When you shut down VS you do get a dispose on the notebook editor, but you do not get a close. In the notebookEditorProvider the disposeResourceRelatedToNotebookEditor is now only called onClose, not on dispose (since dispose is happening more often, such as moving panes).

disposeResourceRelatedToNotebookEditor basically does the following:
run closeEditor in notebookEditorProvider
dispose editor in notebookEditor
dispose model in baseModel

In the shutdown case the model is disposed via the notebookStorageProvider which puts all models on the dispose queue. In the notebookEditorProvider onEditorOpened we hook up to run the closedEditor function when the editor is closed. And the notebookEditor is set to call the .closed event when the model is disposed. Meaning that on shutdown the following happens:

dispose model in baseModel
disposing model causes closed to file in notebookEditor
run closeEditor in notebookEditorProvider

So basically the exact same things happen here. However I did note that notebookEditor dispose just fires the closed event. But it only hooks up firing the closed event to when the model is disposed not the full dispose. What I have changed this to now is functionally the same, since the only thing that dispose is doing is firing the closed event. But I think this makes it a bit safer in the future if we want to add more to the dispose here.

Comment (IanMatthewHuff) [reply src/client/datascience/notebook/notebookEditor.ts:100]: @DonJayamanne @rchiodo @joyceerhl Made a tweak on the review. Given that I'm planning on porting this to release wanted to triple check on signoff on it.

**Review by rchiodo — APPROVED** [review]

**Review by DavidKutu — APPROVED** [review]

**Comment by IanMatthewHuff** [comment]
Test failures due to insiders API update.
